src/renderer/base/renderer.cpp

```diff
@@ -772,10 +783,38 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
             THROW_IF_FAILED(pEngine->PaintBufferLine({ clusters.data(), clusters.size() }, screenPoint, trimLeft, lineWrapped));
 
             // If we're allowed to do grid drawing, draw that now too (since it will be coupled with the color data)
+            // We're only allowed to draw the grid lines under certain circumstances.
             if (_pData->IsGridLineDrawingAllowed())
             {
-                // We're only allowed to draw the grid lines under certain circumstances.
-                _PaintBufferOutputGridLineHelper(pEngine, currentRunColor, cols, screenPoint);
+                // See GH: 803
+                // If we found a wide character while we looped above, it's possible we skipped over the right half
+                // attribute that could have contained different line information than the left half.
+                if (containsWideCharacter)
+                {
+                    // Start from the original position.
+                    auto lineIt = originalIt;
+                    // Start from the original target.
+                    auto lineTarget = target;

```

Comment:
I think I've figured out where's it's going wrong, We're resetting the iterator and target here every time we render a section of the line. So when there are multiple sections, only the first part gets rendered correctly. The only way I can see this working is if we do the entire grid rendering in one go outside the main while loop. But that means you lose the non-widechar optimization, unless you can calculate that up front somehow.

Reply:
Nnnn. Thanks for digging. I'll check this against your test code on Friday when I'm back to work.

Reply:
Friday I learned I needed a new roof so I took the rest of the day to rage at that. Going to peek in what's left of today and/or Wednesday now. Thanks.

Reply:
Hmmmmm I'm rooting through this and as I patch it up and patch it up to make it happen inside the same loop, it's only getting more and more complicated. I think I either need to

1. Make the line drawing a separate pass through the iterators.
2. Make this pass through the loop, that is gathering up `Cluster` as a rendering unit, also gather up some sort of `LineCluster` units. And then dispatch the two of those later.

The only reason I'd lean toward the latter is because walking the iterators isn't super cheap, as of my memory from last time I profiled it. 

Additionally, I need to check what the v1 console does here with your sample code. I was under the presumption that one could set a different set of lines to the right half of the character than the left half, for whatever reason. If that wasn't possible in the v1 console and it did something like, say, only use the left-half lines for drawing around the entire 2-col character... then I'd like to change this back to do the same thing.

Reply:
FWIW, my attempt to solve this was using the second approach, i.e. building up a cluster of grid lines for each pass of the main loop. What I was storing was the result of the `s_GetGridlines` call rather than the original `TextAttribute`, since I thought that would be a little more efficient. The actual grid color you should be able to get from the `currentRunColor` for each pass. 

That said, I'm not sure I had all the edge cases correct (like backing up when you have a trailing DBCS char), and I actually thought your iterator approach was much nicer since it avoided the extra memory allocations needed for the list.

Reply:
2 days of not looking at it and one gentle nudge got my change working. So I'm resolving this.